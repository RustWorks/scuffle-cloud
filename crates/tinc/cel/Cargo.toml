[package]
name = "tinc-cel"
version = "0.1.0"
edition = "2024"
description = "A CEL Type used by tinc-build and tinc for validating protobuf messages"
repository = "https://github.com/scufflecloud/scuffle"
authors = ["Scuffle <opensource@scuffle.cloud>"]
readme = "README.md"
documentation = "https://docs.rs/tinc-build"
license = "MIT OR Apache-2.0"
keywords = ["grpc", "protobuf", "tonic", "codegen"]

[dependencies]
num-traits = "0.2.19"
bytes = "1.5"
float-cmp = "0.10"
thiserror = "2"
chrono = "0.4"
regex = "1"
linkme = "0.3"
fmtools = "0.1"
uuid = "1"
url = "2"
email_address = "0.2.9"
scuffle-workspace-hack.workspace = true

[features]
runtime = []

