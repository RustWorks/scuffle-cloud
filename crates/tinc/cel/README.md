<!-- cargo-sync-rdme title [[ -->
# tinc-cel
<!-- cargo-sync-rdme ]] -->

> [!WARNING]  
> This crate is under active development and may not be stable.

<!-- cargo-sync-rdme badge [[ -->
![License: MIT OR Apache-2.0](https://img.shields.io/crates/l/tinc-cel.svg?style=flat-square)
[![docs.rs](https://img.shields.io/docsrs/tinc-cel.svg?logo=docs.rs&style=flat-square)](https://docs.rs/tinc-cel)
[![crates.io](https://img.shields.io/crates/v/tinc-cel.svg?logo=rust&style=flat-square)](https://crates.io/crates/tinc-cel)
[![GitHub Actions: ci](https://img.shields.io/github/actions/workflow/status/scufflecloud/scuffle/ci.yaml.svg?label=ci&logo=github&style=flat-square)](https://github.com/scufflecloud/scuffle/actions/workflows/ci.yaml)
[![Codecov](https://img.shields.io/codecov/c/github/scufflecloud/scuffle.svg?label=codecov&logo=codecov&style=flat-square)](https://codecov.io/gh/scufflecloud/scuffle)
<!-- cargo-sync-rdme ]] -->

---

<!-- cargo-sync-rdme rustdoc [[ -->
Currently this is a fully private api used by `tinc` and `tinc-build` to
compile and execute [CEL](https://cel.dev/) expressions.
Check out the [changelog](./CHANGELOG.md).

### Feature flags

* **`runtime`** —  Enables runtime enum type indexing
* **`docs`** —  Enables changelog and documentation of feature flags

### License

This project is licensed under the MIT or Apache-2.0 license.
You can choose between one of them if you use this work.

`SPDX-License-Identifier: MIT OR Apache-2.0`
<!-- cargo-sync-rdme ]] -->
