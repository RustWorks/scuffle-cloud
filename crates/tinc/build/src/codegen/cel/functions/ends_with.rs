use cel_interpreter::{ExecutionError, FunctionContext};
use syn::parse_quote;

use super::Function;
use crate::codegen::cel::compiler::{CompileError, CompiledExpr, CompilerCtx};
use crate::codegen::cel::types::CelType;
use crate::types::{ProtoType, ProtoValueType};

pub struct EndsWith;

// this.stratsWith(arg) -> arg in this
impl Function for EndsWith {
    const NAME: &'static str = "endsWith";

    fn compile(ctx: CompilerCtx) -> Result<CompiledExpr, CompileError> {
        let Some(this) = &ctx.this else {
            return Err(CompileError::MissingTarget {
                func: Self::NAME,
                message: "this is required when calling the endsWith function".to_string(),
            });
        };

        if ctx.args.len() != 1 {
            return Err(CompileError::InvalidFunctionArgumentCount {
                func: Self::NAME,
                expected: 1,
                got: ctx.args.len(),
            });
        }

        let arg = ctx.resolve(&ctx.args[0])?;
        if !arg.ty.can_be_cel() {
            return Err(CompileError::TypeConversion {
                ty: Box::new(arg.ty.clone()),
                message: "the endsWith function can only be called with CEL value argument types".to_string(),
            });
        }

        if !this.ty.can_be_cel() {
            return Err(CompileError::TypeConversion {
                ty: Box::new(this.ty.clone()),
                message: "the endsWith function can only be called with CEL value argument types".to_string(),
            });
        }

        Ok(CompiledExpr {
            expr: parse_quote! {
                ::tinc::__private::cel::CelValue::cel_ends_with(
                    #this,
                    #arg,
                )?
            },
            ty: CelType::Proto(ProtoType::Value(ProtoValueType::Bool)),
        })
    }

    fn interpret(fctx: &FunctionContext) -> Result<cel_interpreter::Value, ExecutionError> {
        let Some(this) = &fctx.this else {
            return Err(ExecutionError::missing_argument_or_target());
        };

        if fctx.args.len() != 1 {
            return Err(ExecutionError::invalid_argument_count(1, fctx.args.len()));
        }

        let arg = fctx.ptx.resolve(&fctx.args[0])?;

        match (this, arg) {
            (cel_interpreter::Value::String(s), cel_interpreter::Value::String(t)) => {
                Ok(cel_interpreter::Value::Bool(s.ends_with(t.as_str())))
            }
            (cel_interpreter::Value::Bytes(s), cel_interpreter::Value::Bytes(t)) => {
                Ok(cel_interpreter::Value::Bool(s.ends_with(t.as_slice())))
            }
            _ => Ok(cel_interpreter::Value::Bool(false)),
        }
    }
}
