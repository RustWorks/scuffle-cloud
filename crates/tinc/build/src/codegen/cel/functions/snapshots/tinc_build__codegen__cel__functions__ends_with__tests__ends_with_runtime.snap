---
source: crates/tinc/build/src/codegen/cel/functions/ends_with.rs
expression: compiled
---
exit status: 0
--- stdout 
#![feature(prelude_import)]
#[prelude_import]
use std::prelude::rust_2024::*;
#[macro_use]
extern crate std;
#[allow(dead_code)]
fn string_ends_with(
    input: &std::string::String,
) -> Result<bool, ::tinc::__private::cel::CelError<'_>> {
    Ok(
        ::tinc::__private::cel::CelValue::cel_ends_with(
            ::tinc::__private::cel::CelValueConv::conv(input),
            ::tinc::__private::cel::CelValue::String(
                ::tinc::__private::cel::CelString::Borrowed("ing"),
            ),
        )?,
    )
}
